src/sentry/services/hybrid_cloud/auth/impl.py

```diff
@@ -240,6 +257,64 @@ def get_org_ids_with_scim(
     def get_auth_providers(self, organization_id: int) -> List[ApiAuthProvider]:
         return list(AuthProvider.objects.filter(organization_id=organization_id))
 
+    def handle_new_membership(
+        self,
+        request: Request,
+        organization: ApiOrganization,
+        auth_identity: ApiAuthIdentity,
+        auth_provider: ApiAuthProvider,
+    ) -> Tuple[APIUser, ApiOrganizationMember | None]:
+        # TODO: Might be able to keep hold of the APIUser object whose ID was
+        #  originally passed to construct the ApiAuthIdentity object
+        user = User.objects.get(id=auth_identity.user_id)
+        serial_user = user_service.serialize_user(user)
+
+        # This raises an AvailabilityError
+        # TODO: Extract ApiInviteHelper methods into new service methods
+        orm_org = Organization.objects.get(id=organization.id)
+
+        # If the user is either currently *pending* invite acceptance (as indicated
+        # from the invite token and member id in the session) OR an existing invite exists on this
+        # organization for the email provided by the identity provider.
+        invite_helper = ApiInviteHelper.from_session_or_email(
+            request=request, organization=orm_org, email=user.email
+        )
+
+        # If we are able to accept an existing invite for the user for this
+        # organization, do so, otherwise handle new membership
+        if invite_helper:
+            if invite_helper.invite_approved:
+                om = invite_helper.accept_invite(user)
+                return serial_user, DatabaseBackedOrganizationService.serialize_member(om)
+
+            # It's possible the user has an _invite request_ that hasn't been approved yet,
+            # and is able to join the organization without an invite through the SSO flow.
+            # In that case, delete the invite request and create a new membership.
+            invite_helper.handle_invite_not_approved()
+
+        flags = ApiOrganizationMemberFlags(sso__linked=True)
+        # if the org doesn't have the ability to add members then anyone who got added
+        # this way should be disabled until the org upgrades
+        if not features.has("organizations:invite-members", orm_org):
+            flags.member_limit__restricted = True
+
+        # Otherwise create a new membership
+        om = organization_service.add_organization_member(
```

Comment:
Cool.  In the future, this will become simpler: we'll create an `OrganizationMemberMapping` orm locally on the control, and that will synchronize back.  But this is sufficient for current world compatibility.